quartz/components/ExplorerNode.tsx

```diff
@@ -0,0 +1,205 @@
+  }
+
+  // Add new file to tree
+  add(file: Data, splice: number = 0) {
+    this.insert({ file, path: file.filePath!.split("/").splice(splice) }, -1)
+  }
+
+  // Print tree structure (for debugging)
+  print(depth: number = 0) {
+    let folderChar = ""
+    if (!this.file) folderChar = "|"
+    console.log("-".repeat(depth), folderChar, this.name, this.depth)
+    this.children.forEach((e) => e.print(depth + 1))
+  }
+
+  /**
+   * Get folder representation with state of tree.
+   * Intended to only be called on root node before changes to the tree are made
+   * @param collapsed default state of folders (collapsed by default or not)
+   * @returns array containing folder state for tree
+   */
+  getFolderPaths(collapsed: boolean): FolderState[] {
+    const folderPaths: FolderState[] = []
+
+    const traverse = (node: FileNode, currentPath: string) => {
+      if (!node.file) {
+        const folderPath = currentPath + (currentPath ? "/" : "") + node.name
+        if (folderPath !== "") {
+          folderPaths.push({ path: folderPath, collapsed })
+        }
+        node.children.forEach((child) => traverse(child, folderPath))
+      }
+    }
+
+    traverse(this, "")
+
+    return folderPaths
+  }
+
+  // Sort order: folders first, then files. Sort folders and files alphabetically
+  sort() {
+    this.children = this.children.sort((a, b) => {
+      if ((!a.file && !b.file) || (a.file && b.file)) {
+        return a.name.localeCompare(b.name)
+      }
+      if (a.file && !b.file) {
+        return 1
+      } else {
+        return -1
+      }
+    })
+
+    this.children.forEach((e) => e.sort())
+  }
+}
+
+type ExplorerNodeProps = {
+  node: FileNode
+  opts: Options
+  fullPath?: string
+}
+
+export function ExplorerNode({ node, opts, fullPath }: ExplorerNodeProps) {
+  // Get options
+  const folderBehavior = opts.folderClickBehavior
+  const collapseFolders = opts.folderDefaultState === "collapsed"
+
+  // Calculate current folderPath
+  let pathOld = fullPath ? fullPath : ""
+  let folderPath = ""
+  if (node.name !== "") {
+    folderPath = `${pathOld}/${node.name}`
+  }
+
+  // Helpers to disable pointer-events and opacity for all appropriate elements
+  const isInvisInner = node.depth > 1 && collapseFolders
+  const isInvisOuter = node.depth > 0 && collapseFolders
+
+  return (
+    <div class={`${isInvisOuter ? "no-pointer" : ""}`}>
+      {node.file ? (
+        // Single file node
+        <li key={node.file.slug} class="no-pointer">
+          <a
+            href={`/${node.file.slug}`}
+            data-for={node.file.slug}
+            class={`clickable ${isInvisInner ? "no-pointer" : ""}`}
+          >
+            {node.file.frontmatter?.title}
+          </a>
+        </li>
+      ) : (
+        <div>
+          {node.name !== "" && (
+            // Node with entire folder
+            // Render svg button + folder name, then children
+            <div class={`folder-container ${isInvisInner ? "no-pointer" : ""}`}>
```

Comment:
i wonder if we should just make a component for a collapsable list lol

Reply:
yes, it would make life so much easier, we should probably do that after merging this in a seperate PR. I found a really clean way to do this with a grid based approach (example [here](https://www.stefanjudis.com/snippets/how-to-animate-height-with-css-grid/)) and am currently working on refactoring the file explorer to use this approach. saves so much jank with setting pointer-events: none, makes life easier and we should use it everywhere (no more setting heights manually, etc)

Reply:
that would be super epic, excited to see what you come up with!

Reply:
fixed this for explorer for now, could do this for the rest of quartz in the future in a separate issue/PR